cpp/src/arrow/ipc/metadata.cc

```diff
@@ -591,7 +593,20 @@ static Status SchemaToFlatbuffer(FBB& fbb, const Schema& schema,
     field_offsets.push_back(offset);
   }
 
-  *out = flatbuf::CreateSchema(fbb, endianness(), fbb.CreateVector(field_offsets));
+  /// Custom metadata
+  auto custom_metadata_ = schema.custom_metadata();
+  std::vector<KeyValueOffset> key_value_offsets;
+  size_t metadata_size = custom_metadata_.size();
+  key_value_offsets.reserve(metadata_size);
+  for (size_t i = 0; i < metadata_size; ++i) {
+    auto key = custom_metadata_.key(i);
+    auto value = custom_metadata_.value(i);
```

Comment:
NB. `auto` here will invoke the string copy constructor. You have to use `const auto&`